extensions/rich_text_components/Video/directives/oppia-noninteractive-video.component.ts

```diff
@@ -0,0 +1,121 @@
+
+let apiLoaded = false;
+@Component({
+  selector: 'oppia-noninteractive-video',
+  templateUrl: './video.component.html',
+  styleUrls: []
+})
+export class NoninteractiveVideo implements OnInit {
+  @Input() autoplayWithValue: string;
+  @Input() endWithValue: string;
+  @Input() startWithValue: string;
+  @Input() videoIdWithValue: string;
+
+  playerVars = {
+    autoplay: 0,
+    origin: 'http://localhost:8181'
+  };
+  start: number;
+  end: number;
+  tabIndexVal: number;
+  videoId: string;
+  width: number;
+
+  constructor(
+    private autoplayedVideosService: AutoplayedVideosService,
+    private contextService: ContextService,
+    private el: ElementRef,
+    private htmlEscaperService: HtmlEscaperService
+  ) {}
+
+  ngOnInit(): void {
+    if (!apiLoaded) {
+      // This code loads the IFrame Player API code asynchronously, according to
+      // the instructions at
+      // https://developers.google.com/youtube/iframe_api_reference#Getting_Started
+      const tag = document.createElement('script');
+      tag.src = 'https://www.youtube.com/iframe_api';
+      document.body.appendChild(tag);
+      apiLoaded = true;
+    }
+    const start = this.htmlEscaperService.escapedJsonToObj(
+      this.startWithValue) as string;
+    const end = this.htmlEscaperService.escapedJsonToObj(
+      this.endWithValue) as string;
+    this.start = +start;
+    this.end = +end;
+    this.videoId = this.htmlEscaperService.escapedJsonToObj(
+      this.videoIdWithValue) as string;
+
+    // Check whether creator wants to autoplay this video or not.
+    const autoplayVal = this.htmlEscaperService.escapedJsonToObj(
+      this.autoplayWithValue);
+
+    // This code helps in visibility of video. It checks whether
+    // mid point of video frame is in the view or not.
+    const rect = this.el.nativeElement.getBoundingClientRect();
+    const clientHeight = window.innerHeight;
+    const clientWidth = window.innerWidth;
+    this.width = this.el.nativeElement.width;
+    const isVisible = (
+      (rect.left + rect.right) / 2 < clientWidth &&
+      (rect.top + rect.bottom) / 2 < clientHeight) &&
+      (rect.left > 0 && rect.right > 0);
+
+    // Autoplay if user is in learner view and creator has specified
+    // to autoplay given video.
+    if (this.contextService.getPageContext() ===
+      ServicesConstants.PAGE_CONTEXT.EXPLORATION_PLAYER && autoplayVal) {
+      // If it has been autoplayed then do not autoplay again.
+      if (
+        !this.autoplayedVideosService.hasVideoBeenAutoplayed(
+          this.videoId) && isVisible) {
+        this.playerVars.autoplay = 1;
+        this.autoplayedVideosService.addAutoplayedVideo(this.videoId);
+      }
+    }
+    // (^)Here timeout is set to 900ms. This is time it takes to bring
+    // the frame to correct point in browser and bring user to the main
+    // content. Smaller delay causes checks to be performed even before
+    // the player displays the content of the new card.
```

Comment:
Where?

Reply:
Apparently no where 😂.. Removed this comment. This is now the headache that the angular team has to face with the youtube player component. Not us. :)